File: LeetCode/Medium/0092-reverse-linked-list-ii/0092-reverse-linked-list-ii.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func reverseBetween(head *ListNode, left int, right int) *ListNode {
    if left == right { return head }
    
    dummy := &ListNode{}
    pre := dummy
    dummy.Next = head
    
    for i := 1; i < left; i++ { pre = pre.Next }
    
    curr := pre.Next
    var prev *ListNode
    
    for i := 0; i < right-left+1; i++ {
        temp := curr.Next
        curr.Next = prev
        prev = curr
        curr = temp
    }
    
    pre.Next.Next = curr
    pre.Next = prev
    
    return dummy.Next
}

File: LeetCode/Medium/0092-reverse-linked-list-ii/0092-reverse-linked-list-ii.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseBetween(ListNode* head, int left, int right) {
        if (left == right) return head;
        
        ListNode* dummy = new ListNode();
        dummy->next = head;
        ListNode* pre = dummy;
        
        for (int i = 1; i < left; ++i) pre = pre->next;
        
        ListNode* curr = pre->next;
        ListNode* prev = nullptr;
        
        for (int i = 0; i < right-left+1; ++i){
            ListNode* temp = curr->next;
            curr->next = prev;
            prev = curr;
            curr = temp;
        }
        
        pre->next->next = curr;
        pre->next = prev;
        
        return dummy->next;
    }
};